### PR TITLE
fix: secure built-in filesystem MCP root handling

### LINKED ISSUE #12447: [Bug]: Built-in MCP service filesystem can access arbitrary directories
> [!NOTE]
> This issue was translated by Claude.

### Issue Checklist

- [x] I understand that issues are for feedback and problem solving, not for complaining in the comment section, and will provide as much information as possible to help solve the problem.
- [x] My issue is not listed in the [FAQ](https://github.com/CherryHQ/cherry-studio/issues/3881).
- [x] I've looked at **pinned issues** and searched for existing [Open Issues](https://github.com/CherryHQ/cherry-studio/issues), [Closed Issues](https://github.com/CherryHQ/cherry-studio/issues?q=is%3Aissue%20state%3Aclosed), and [Discussions](https://github.com/CherryHQ/cherry-studio/discussions), no similar issue or discussion was found.
- [x] I've filled in short, clear headings so that developers can quickly identify a rough idea of what to expect when flipping through the list of issues. And not "a suggestion", "stuck", etc.
- [x] I've confirmed that I am using the latest version of Cherry Studio.

### Platform

Windows

### Version

v1.7.8

### Software
The built-in filesystem is version 2.0.0 and starts normally.
Image

### Problem
The configured directory is: `E:\\test` (`E:/test` was also tried)
However, other directories can still be accessed, as shown in the image below. It appears that all permissions are automatically approved by default, including deletion and editing. There is a high risk of accidentally deleting files due to misuse.
Image

### Steps To Reproduce

1. Install the built-in filesystem
2. On the first startup, it reports `Error invoking remote method 'mcp:list-tools': McpError: MCP error -32601: Method not Found`, second startup is normal
3. Use the model to call the MCP service to access folders outside the configured directory

### Expected Behavior

1. Proper permission settings
2. Sensitive actions: auto-approval for editing, modifying, and deleting should be disabled by default.

### Relevant Log Output

```shell

```

### Additional Context

_No response_

---
<details>
<summary>Original Content</summary>

### Issue Checklist

- [x] I understand that issues are for feedback and problem solving, not for complaining in the comment section, and will provide as much information as possible to help solve the problem.
- [x] My issue is not listed in the [FAQ](https://github.com/CherryHQ/cherry-studio/issues/3881).
- [x] I've looked at **pinned issues** and searched for existing [Open Issues](https://github.com/CherryHQ/cherry-studio/issues), [Closed Issues](https://github.com/CherryHQ/cherry-studio/issues?q=is%3Aissue%20state%3Aclosed), and [Discussions](https://github.com/CherryHQ/cherry-studio/discussions), no similar issue or discussion was found.
- [x] I've filled in short, clear headings so that developers can quickly identify a rough idea of what to expect when flipping through the list of issues. And not "a suggestion", "stuck", etc.
- [x] I've confirmed that I am using the latest version of Cherry Studio.

### Platform

Windows

### Version

v1.7.8

### 软件
内置的filesystem为2.0.0，正常启动。
Image

### 问题
设置的目录为：`E:\\test`（`E:/test`也试过）
但是仍然可以访问其他目录，如下图。此外，默认所有权限都是自动批准状态，包括删除、编辑，误用删除文件风险大。
Image

### Steps To Reproduce

1. 安装内置的filesystem
2. 第一次启动会报`Error invoking remote method 'mcp:list-tools': McpError: MCP error -32601: Method not Found`，第二次正常启动
3. 通过模型调用该mcp服务访问设置目录外的文件夹

### Expected Behavior

1. 正确的权限设置
2. 敏感行为：如编辑、修改、删除自动批准功能默认不启用。

### Relevant Log Output

```shell

```

### Additional Context

_No response_
</details>

### PR DESCRIPTION
**Pull request by DeJeune**
### What this PR does

Before this PR:
- The built-in filesystem MCP server accepted absolute paths outside the configured workspace root.
- The in-memory filesystem server only read `WORKSPACE_ROOT`, so roots configured via `args` were ignored.
- The server registered tool handlers asynchronously, which could cause first-load `mcp:list-tools` failures.
- Sensitive filesystem tools could be auto-approved by default.

image

After this PR:
- Filesystem MCP paths are constrained to the configured workspace root after realpath resolution.
- The built-in filesystem server resolves its root from `WORKSPACE_ROOT` or the configured `args`.
- Tool handlers are registered before async directory initialization starts.
- Built-in filesystem `write`, `edit`, and `delete` require manual approval by default, including a backfill for existing configs.

image

Fixes #12447

### Why we need it and why it was done in this way

The following tradeoffs were made:
- Existing filesystem MCP configurations now default to safer approval behavior for destructive tools.
- Path validation resolves symlinks and nearest existing ancestors to block directory escapes for both existing and newly created paths.

The following alternatives were considered:
- Trusting UI configuration alone without enforcing the root in the filesystem tool layer.
- Keeping auto-approve defaults unchanged and relying on users to disable risky tools manually.
- Waiting for async initialization to finish before serving requests instead of registering handlers immediately.

Links to places where the discussion took place: #12447, #13280

### Breaking changes

The built-in filesystem MCP server now requires manual approval for `write`, `edit`, and `delete` by default. Users who relied on automatic approval for these tools need to re-enable auto-approve explicitly.

### Special notes for your reviewer

- This PR does not change Redux state shape or IndexedDB schema.
- Added regression coverage for root resolution, path escapes, symlink escapes, and filesystem approval defaults.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
action required: The built-in filesystem MCP server now enforces the configured workspace root and requires manual approval for write, edit, and delete operations by default.
```

### CODE REVIEW

File: src/main/mcpServers/filesystem/types.ts [hunk 1]
```diff
@@ -39,27 +39,59 @@ export function expandHome(filepath: string): string {
...
+        const relativeSuffix = path.relative(currentPath, targetPath)
+        return normalizePath(path.join(realCurrentPath, relativeSuffix))
+      } catch {
+        const parentPath = path.dirname(currentPath)
+        if (parentPath === currentPath) {
+          return normalizePath(targetPath)
+        }
+        currentPath = parentPath
+      }
+    }
```
Comment (EurFelux) [inline src/main/mcpServers/filesystem/types.ts:66]: **Potential edge case: root with no existing ancestor**

When the workspace root path itself doesn't exist and neither does any of its ancestors (very extreme edge case, e.g. an external drive path before mounting), this loop reaches the filesystem root and falls back to `normalizePath(targetPath)` without resolving symlinks.

In practice this is unlikely to be exploitable since both `resolvedRoot` and `resolvedPath` would both get this fallback treatment, but worth considering adding a log warning when the loop exhausts all ancestors.

```ts
if (parentPath === currentPath) {
  logger.warn('Could not resolve any existing ancestor for path', { targetPath })
  return normalizePath(targetPath)
}
```

nit: Not a blocker.

File: src/main/mcpServers/filesystem/types.ts [hunk 2]
```diff
@@ -39,27 +39,59 @@ export function expandHome(filepath: string): string {
...
-    } catch {
-      return normalizePath(absolute)
-    }
+  const resolvedRoot = await resolveRealOrNearestExistingPath(path.resolve(root))
+  const resolvedPath = await resolveRealOrNearestExistingPath(absolute)
+
+  if (!isPathWithinRoot(resolvedPath, resolvedRoot)) {
+    throw new Error(`Access denied: Path is outside the configured workspace root: ${requestedPath}`)
   }
+
```
Comment (EurFelux) [inline src/main/mcpServers/filesystem/types.ts:94]: **TOCTOU consideration (informational)**

There's an inherent TOCTOU (Time-of-check-time-of-use) window between `validatePath` resolving the real path and the actual file operation. For example, a symlink could be created between the validation and the file read/write.

This is a known limitation of path-based security validation in general and not specific to this PR. Just flagging for awareness — mitigating this fully would require something like `O_NOFOLLOW` or `openat()` which is beyond scope here.

Not a blocker — the current approach significantly raises the bar compared to no validation at all.

File: src/main/mcpServers/filesystem/__tests__/filesystem.test.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import { validatePath } from '../types'
+
+describe('filesystem MCP security', () => {
+  const tempDirs: string[] = []
+
+  async function createTempDir(prefix: string) {
+    const tempRoot = path.join(process.cwd(), '.context', 'vitest-temp')
+    await fs.mkdir(tempRoot, { recursive: true })
+    const tempDir = await fs.mkdtemp(path.join(tempRoot, prefix))
+    tempDirs.push(tempDir)
```
Comment (EurFelux) [inline src/main/mcpServers/filesystem/__tests__/filesystem.test.ts:18]: **nit: Consider using `os.tmpdir()` instead of project-local temp dirs**

Using `process.cwd()/.context/vitest-temp` creates artifacts inside the project directory. This could accidentally be committed or interfere with other tooling. Using `os.tmpdir()` is more conventional and avoids polluting the working directory.

```ts
const tempRoot = path.join(os.tmpdir(), 'cherry-studio-vitest')
```

Also, is `.context` in `.gitignore`? If not, these temp files could be accidentally committed.

File: src/main/mcpServers/filesystem/__tests__/filesystem.test.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  afterEach(async () => {
+    await Promise.all(tempDirs.splice(0).map((tempDir) => fs.rm(tempDir, { recursive: true, force: true })))
+  })
+
+  it('prefers WORKSPACE_ROOT and falls back to args for filesystem root', () => {
+    expect(resolveFilesystemBaseDir(['C:/args-root'], {})).toBe('C:/args-root')
+    expect(resolveFilesystemBaseDir(['C:/args-root'], { WORKSPACE_ROOT: 'C:/env-root' })).toBe('C:/env-root')
+    expect(resolveFilesystemBaseDir([], {})).toBeUndefined()
+  })
```
Comment (EurFelux) [inline src/main/mcpServers/filesystem/__tests__/filesystem.test.ts:31]: **Suggestion: Add test cases for relative paths and `~` expansion**

The current tests cover the key security scenarios well (root containment, symlink escapes). Consider adding coverage for:

1. **Relative path resolution**: `validatePath('../../etc/passwd', workspaceRoot)` should be rejected
2. **`~` expansion**: `validatePath('~/sensitive-file', workspaceRoot)` should be rejected if home dir is outside root
3. **`validatePath` with no `baseDir`**: verify fallback to `process.cwd()` still enforces containment

These would strengthen the regression suite for path traversal attacks.

File: src/renderer/src/store/mcp.ts [hunk 5]
```diff
@@ -240,6 +242,16 @@ export const builtinMCPServers: BuiltinMCPServer[] = [
...
  */
 export const initializeMCPServers = (existingServers: MCPServer[], dispatch: (action: any) => void): void => {
+  const existingFilesystemServer = existingServers.find((server) => server.name === BuiltinMCPServerNames.filesystem)
+  if (existingFilesystemServer && existingFilesystemServer.disabledAutoApproveTools === undefined) {
+    dispatch(
+      updateMCPServer({
+        ...existingFilesystemServer,
+        disabledAutoApproveTools: [...filesystemManualApprovalTools]
+      })
+    )
```
Comment (EurFelux) [inline src/renderer/src/store/mcp.ts]: **Question: Backfill runs on every app initialization**

This backfill code runs on every `initializeMCPServers` call. After the first run, `disabledAutoApproveTools` will be set, so the `=== undefined` check will prevent re-applying.

However, what if a user **intentionally** clears `disabledAutoApproveTools` (sets it to `undefined` or deletes the property)? The next initialization would re-apply the defaults, overriding their choice.

A more robust approach might be to use a separate one-time migration flag, e.g. `_filesystemApprovalMigrated: true`, to track whether the backfill has been applied. But this is a minor concern — the current approach is pragmatic and covers the common case well.

nit: Not a blocker.

**Comment by EurFelux** [comment]
> [!NOTE]
> This issue/comment/review was translated by Claude.

I recall that this issue didn't exist earlier. It may have been introduced by #11937 without strict review.

---
<details>
<summary>Original Content</summary>

印象里早些时候是没有这个问题的。可能是 #11937 未经严格审查引入了该问题。

</details>

File: src/main/mcpServers/filesystem/types.ts (hunk 1 above)
Comment (DeJeune) [reply src/main/mcpServers/filesystem/types.ts:66]: Added a warning when ancestor resolution exhausts the path tree before finding any existing entry. This keeps the existing fallback behavior but makes the edge case visible in logs.

File: src/main/mcpServers/filesystem/types.ts (hunk 2 above)
Comment (DeJeune) [reply src/main/mcpServers/filesystem/types.ts:94]: Added regression coverage for relative traversal (`../outside.txt`), `~` expansion outside the root, and the `process.cwd()` fallback path when `baseDir` is omitted.

File: src/renderer/src/store/mcp.ts (hunk 5 above)
Comment (DeJeune) [reply src/renderer/src/store/mcp.ts]: I kept the current backfill approach for now. Adding a dedicated migration flag would mean introducing more state into `src/renderer/src/store/mcp.ts`, which is currently v2-blocked. In this case I think treating `undefined` as “apply the safer default” is the better tradeoff until a broader migration path is in place.

**Comment by vaayne** [comment]
> [!NOTE]
> This issue/comment/review was translated by Claude.

@kangfenmao mentioned this issue during testing before. I intentionally didn't have this restriction because some people configured directories but had more issues, but adding it is also good.

---
<details>
<summary>Original Content</summary>

之前 @kangfenmao 测试的时候说过这个问题，我是故意没有这个限制的，因为有人配置了目录但是问题更多，不过加上了也好

</details>

**Comment by vaayne** [comment]
> [!NOTE]
> This issue/comment/review was translated by Claude.

I still think using absolute paths would be better, it's up to you guys, I'm fine either way

---
<details>
<summary>Original Content</summary>

但是我仍然觉得使用绝对路径会更好，看你们吧，和不和我都可以

</details>

**Review by EurFelux — APPROVED** [review]
LGTM! Solid security hardening for the built-in filesystem MCP server.

**What's well done:**
- `resolveRealOrNearestExistingPath` correctly handles both existing files (via `realpath`) and new files (nearest ancestor resolution), blocking symlink-based directory escapes
- Splitting `initialize()` → sync `registerHandlers()` + async `ensureBaseDir()` fixes the tool registration race condition
- `isPathWithinRoot` with `path.relative` is clean and handles edge cases (Windows case-insensitivity, cross-drive paths)
- Good test coverage: root containment, symlinks, relative traversal, `~` expansion, `cwd()` fallback
- Backfill approach for `disabledAutoApproveTools` is pragmatic given v2 constraints

All previous review suggestions have been addressed. Approving.

File: src/renderer/src/store/mcp.ts [hunk 6]
```diff
@@ -240,6 +242,16 @@ export const builtinMCPServers: BuiltinMCPServer[] = [
  * @param dispatch Redux dispatch function
  */
 export const initializeMCPServers = (existingServers: MCPServer[], dispatch: (action: any) => void): void => {
+  const existingFilesystemServer = existingServers.find((server) => server.name === BuiltinMCPServerNames.filesystem)
+  if (existingFilesystemServer && existingFilesystemServer.disabledAutoApproveTools === undefined) {
```
Comment (zhibisora) [inline src/renderer/src/store/mcp.ts]: **Note:** This backfill path appears to be unreachable in production. I couldn't find any runtime call site for , while the app reads MCP server configs directly from persisted Redux state. That means existing filesystem server entries keep the old auto-approve behavior after upgrade, so this only affects newly added servers. If that grandfathering is intentional, please document it explicitly.

Comment (DeJeune) [reply src/renderer/src/store/mcp.ts]: Good catch — you're right, `initializeMCPServers` is never called at runtime. The backfill was dead code.

Fixed by moving the backfill into the redux-persist migration system as version 200, which is the mechanism that actually runs against persisted state on app startup. The dead backfill code has been removed from `initializeMCPServers`, and the tests now verify the migration path directly.

File: src/main/mcpServers/filesystem/config.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+export function resolveFilesystemBaseDir(args: string[] = [], envs: Record<string, string> = {}): string | undefined {
+  const envWorkspaceRoot = envs.WORKSPACE_ROOT?.trim()
+  if (envWorkspaceRoot) {
+    return envWorkspaceRoot
+  }
+
+  return args.find((arg) => typeof arg === 'string' && arg.trim().length > 0)?.trim()
```
Comment (zhibisora) [inline src/main/mcpServers/filesystem/config.ts:7]: **Note:** This change makes the built-in filesystem server effectively single-root by taking only the first non-empty arg, while the settings UI and copy still imply that multiple filesystem directories can be configured. I do not think this needs to block the security fix, but it would be good to either align the UI/copy with a single-root model or track multi-root support explicitly in a follow-up issue.

File: src/main/mcpServers/filesystem/tools/glob.ts [hunk 8]
```diff
@@ -26,7 +26,7 @@ export const globToolDefinition = {
 - Patterns with "/" (e.g., "src/*.ts") match relative to the search path
 - Pattern syntax: * (any chars), ** (any path), {a,b} (alternatives), ? (single char)
 - Results are limited to 100 files
-- The path parameter must be an absolute path if specified
+- The path parameter must resolve within the configured workspace root if specified
```
Comment (zhibisora) [inline src/main/mcpServers/filesystem/tools/glob.ts:29]: **Security:** This updated description now promises that the path stays within the configured workspace root, but handleGlobTool still runs ripgrep with --follow and never re-validates each returned match against the root. A symlinked directory inside the workspace can therefore still make glob enumerate files outside the configured root. Please either drop --follow or validate each ripgrep result with the same realpath and root check before returning it.

**Comment by zhibisora** [comment]
Follow-up: to clarify my earlier COMMENT review, I consider the remaining `glob` symlink-traversal gap blocking and would like that fixed before merge. The built-in filesystem MCP can still enumerate files outside the configured root by following symlinked directories without re-validating each returned match against the root.

The single-root vs multi-directory configuration note is non-blocking and can be addressed either in this PR or in a follow-up issue.

**Comment by DeJeune** [comment]
Fixed in 61d68c6af. All three enumeration tools now re-validate each result path against the workspace root after resolution:

- **`glob`**: Each file path returned by ripgrep `--follow` is passed through `validatePath()` before inclusion. Symlink escapes are silently dropped.
- **`grep`**: Both the ripgrep code path and the fallback `searchDirectory` now validate paths. For ripgrep results, each match's file path is checked; for the fallback, each traversed entry is validated before descending or reading.
- **`ls`**: In recursive mode, symlinked directories are validated via `validatePath()` before descending into them.

Added two regression tests:
- `glob excludes files reached via symlinked directories outside root`
- `ls excludes symlinked directories outside root in recursive mode`
